lib/fs/inode_manager.cpp

```diff
@@ -0,0 +1,134 @@
+#include "fs.hpp"
+
+INode_Manager::INode_Manager(Fs *fs, u_int64_t block_segment_start,
+                             u_int64_t block_segment_end)
+    : fs(fs), block_segment_start(block_segment_start),
+      block_segment_end(block_segment_end) {
+  max_num_inodes = (block_segment_end - block_segment_start) * INODES_PER_BLOCK;
+}
+
+u_int64_t INode_Manager::get_block_num(u_int64_t inode_num) {
+  if (inode_num > max_num_inodes || inode_num == 0)
+    return 0;
+  u_int64_t block_num =
+      block_segment_start + ((inode_num - 1) / INODES_PER_BLOCK);
+  return block_num;
+}
+u_int64_t INode_Manager::get_block_offset(u_int64_t inode_num) {
+  return ((inode_num - 1) % INODES_PER_BLOCK) * INODE_SIZE;
+}
+
+int INode_Manager::load_inode(INode_Data *inode_data) {
+  char buf[IO_BLOCK_SIZE];
+  int err;
+
+  u_int64_t block_num = get_block_num(inode_data->inode_num);
+  if (block_num == 0)
+    return -1;
+  u_int64_t block_offset = get_block_offset(inode_data->inode_num);
+
+  if ((err = fs->disk->read_block(block_num, buf)) < 0)
+    return err;
+
+  inode_data->deserialize(&buf[block_offset]);
+
+  return 0;
+}
+int INode_Manager::save_inode(INode_Data *inode_data) {
+  char buf[IO_BLOCK_SIZE];

```

Comment:
Inode size is 512 and io block size is 4096. Are we sure that we will not override the data into the 3 following inodes?

Reply:
I don't fully understand, but the function inode_data->serialize should only write < 512 bytes because the number of inode direct datablocks are calculated to keep the inode size < 512.